### PR TITLE
feat: allow options to be passed into puppeteer.defaultArgs

### PR DESCRIPTION
**Pull request by JoelEinbinder**
Default args didn't take launch options, so were missing some args like `--headless` and `--mute-audio` when in headless mode.

Replaces the experimental option `appMode` with an experimental option `setsDefaultViewport`.

### CODE REVIEW

File: lib/Launcher.js [hunk 1]
```diff
@@ -164,7 +131,7 @@ class Launcher {
         connection = Connection.createForPipe(/** @type {!NodeJS.WritableStream} */(chromeProcess.stdio[3]), /** @type {!NodeJS.ReadableStream} */ (chromeProcess.stdio[4]), connectionDelay);
       }
       const ignoreHTTPSErrors = !!options.ignoreHTTPSErrors;
-      const setDefaultViewport = !options.appMode;
+      const {setDefaultViewport = true} = options;
```
Comment (aslushnikov) [inline lib/Launcher.js]: People will appreciate this; let's release this to public api separately.

File: lib/Launcher.js [hunk 2]
```diff
@@ -231,10 +198,32 @@ class Launcher {
   }
 
   /**
+   * @param {!ChromeArgOptions=} options
    * @return {!Array<string>}
    */
-  static defaultArgs() {
-    return DEFAULT_ARGS.concat(AUTOMATION_ARGS);
+  static defaultArgs(options = {}) {
+    const userArgs = options.args || [];
```
Comment (aslushnikov) [inline lib/Launcher.js]: we had `Array.isArray` check before for user args

**Review by aslushnikov — APPROVED** [review]

File: lib/Launcher.js (hunk 2 above)
Comment (JoelEinbinder) [reply lib/Launcher.js]: If someone passes a truthy but non array value, we should throw an error instead of silently moving on.

Comment (aslushnikov) [reply lib/Launcher.js]: I'm all in for an assert

File: test/puppeteer.spec.js [hunk 3]
```diff
@@ -193,13 +194,18 @@ module.exports.addTests = function({testRunner, expect, PROJECT_ROOT, defaultBro
         await browser.close();
       });
       it('should work with no default arguments', async() => {
+        // We need to set a user-data-dir otherwise this test will fail on the windows bots
+        const userDataDir = await mkdtempAsync(TMP_FOLDER);
```
Comment (aslushnikov) [inline test/puppeteer.spec.js]: why this change?

File: lib/Launcher.js [hunk 4]
```diff
@@ -48,62 +48,50 @@ const DEFAULT_ARGS = [
...
-    if (!options.args || !options.args.some(arg => arg.startsWith('--user-data-dir'))) {
-      if (!options.userDataDir)
-        temporaryUserDataDir = await mkdtempAsync(CHROME_PROFILE_PATH);
+    const chromeArguments = !ignoreDefaultArgs ? this.defaultArgs(options) : args;
+    let temporaryUserDataDir = null;
 
-      chromeArguments.push(`--user-data-dir=${options.userDataDir || temporaryUserDataDir}`);
+    if (!chromeArguments.some(argument => argument.startsWith('--remote-debugging-')))
+      chromeArguments.push(pipe ? '--remote-debugging-pipe' : '--remote-debugging-port=0');
+    if (!ignoreDefaultArgs && !chromeArguments.some(arg => arg.startsWith('--user-data-dir'))) {
```
Comment (aslushnikov) [inline lib/Launcher.js]: I don't think we want `ignoreDefaultArgs` to disable temp userDataDir management. We didn't do this before.

File: test/puppeteer.spec.js (hunk 3 above)
Comment (JoelEinbinder) [reply test/puppeteer.spec.js]: Removed.

File: lib/Launcher.js (hunk 4 above)
Comment (JoelEinbinder) [reply lib/Launcher.js]: Fair.

File: lib/Launcher.js [hunk 5]
```diff
@@ -316,28 +322,54 @@ function waitForWSEndpoint(chromeProcess, timeout) {
...
+ * @typedef {Object} ChromeArgOptions
+ * @property {boolean=} headless
+ * @property {Array<string>=} args
+ * @property {string=} userDataDir
+ * @property {boolean=} devtools
+ */
+
 /**
  * @typedef {Object} LaunchOptions
+<<<<<<< HEAD
```
Comment (ebidel) [inline lib/Launcher.js]: uh, what happened here?

Comment (JoelEinbinder) [reply lib/Launcher.js]: earthquake

**Review by aslushnikov — APPROVED** [review]
